### DISCUSSION TITLE
Paginated query trait bound error

### DISCUSSION THREAD

**Discussion by vinicius98s** [diesel-rs/diesel#3199] [Q&A]
Hello, I'm trying to reproduce some examples (with v2.0rc) with this [pagination implementation](https://github.com/diesel-rs/diesel/blob/master/examples/postgres/advanced-blog-cli/src/pagination.rs) but the following code is generating me an error:
```rs
let connection = &mut establish_connection();
let (result, _) = users::table
    .paginate(1)
    .per_page(4)
    .load_and_count_pages::<User>(connection)
    .expect("pagination error");

dbg!(result);
```
```
error[E0277]: the trait bound `table: diesel::query_builder::Query` is not satisfied
  --> src/main.rs:29:10
   |
29 |         .load_and_count_pages::<User>(connection)
   |          ^^^^^^^^^^^^^^^^^^^^ the trait `diesel::query_builder::Query` is not implemented for `table`
   |
note: required because of the requirements on the impl of `diesel::query_builder::Query` for `Paginated<table>`
  --> src/pagination.rs:54:16
   |
54 | impl<T: Query> Query for Paginated<T> {
   |                ^^^^^     ^^^^^^^^^^^^
   = note: required because of the requirements on the impl of `LoadQuery<'_, diesel::PgConnection, (models::User, i64)>` for `Paginated<table>`
note: required by a bound in `Paginated::<T>::load_and_count_pages`
  --> src/pagination.rs:43:15
   |
41 |     pub fn load_and_count_pages<'a, U>(self, conn: &mut PgConnection) -> QueryResult<(Vec<U>, i64)>
   |            -------------------- required by a bound in this
42 |     where
43 |         Self: LoadQuery<'a, PgConnection, (U, i64)>,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Paginated::<T>::load_and_count_pages`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `diesel_demo` due to previous error
```

this is what my `User` struct looks like:
```rs
use crate::schema::users;

#[derive(Selectable, Debug, Queryable)]
#[diesel(table_name = users)]
pub struct User {
    pub id: uuid::Uuid,
    pub username: String,
    pub password: String,
    pub email: String,
    pub photo: Option<String>,
    pub created_at: chrono::NaiveDateTime,
    pub updated_at: chrono::NaiveDateTime,
}
```
but if I select a column for example, it works fine:
```rs
let connection = &mut establish_connection();
let (result, _) = users::table
    .select(users::username)
    .paginate(1)
    .per_page(4)
    .load_and_count_pages::<String>(connection)
    .expect("pagination error");

dbg!(result); // no error, query with usernames
```
Am I missing something? Is it not possible to query the entire `User` like that?

**Comment by weiznich** [comment]
Your question is missing essential details. Please always provide the complete relevant code. In this case this should include your implementation of the `Paginated` type and all related stuff. 

If you literally followed the paginated example shown at the web page, then please keep in mind that this is just an example. That means this implementation is not necessarily complete or perfect. It is just there to demonstrate the general approach. For your query specifically it will likely fail due to the reason that no table type implements `Query` directly. You might need to call `.as_query()` on the table before passing it directly to `.paginate`. Each `QueryDsl` call (so `select(…)` and so on) will handle that internally, so calling any of those messages will prevent such an error from happening.
